passport/index.d.ts

```diff
@@ -55,8 +55,8 @@ declare module 'passport' {
             authenticate(strategy: string|string[], options: AuthenticateOptions, callback?: Function): express.Handler;
             authorize(strategy: string|string[], callback?: Function): express.Handler;
             authorize(strategy: string|string[], options: any, callback?: Function): express.Handler;
-            serializeUser<TUser, TID>(fn: (user: TUser, done: (err: any, id: TID) => void) => void): void;
-            deserializeUser<TUser, TID>(fn: (id: TID, done: (err: any, user: TUser) => void) => void): void;
+            serializeUser<TUser, TID>(fn: (user: TUser, done: (err: any, id?: TID) => void) => void): void;
+            deserializeUser<TUser, TID>(fn: (id: TID, done: (err: any, user?: TUser) => void) => void): void;
```

Comment:
There's two callbacks in the lines you highlighted:

- `fn`: this is a user-provided callback which will be called by the library. this is the kind of callback described in the Handbook section you linked, and it has no optional parameters.
- `done`: this is a *library*-provided callback which will be called by user code. my argument is that this is a different case; it's not the same kind of callback, it's a method of the library which is dynamically provided to user code. and it's perfectly fine to have optional parameters on functions provided by libraries.

a well-known example of this is Express middleware – middleware functions accept a `next` callback [provided by Express](https://github.com/expressjs/express/blob/master/lib/router/index.js#L178) that is called with no arguments to indicate “ok, invoke the next handler” or can be called with an error to indicate failure.

and those functions are [typed in DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/express-serve-static-core/index.d.ts#L22) with `err` as an optional parameter:

```ts
interface NextFunction {
    (err?: any): void;
}
```

I appreciate the effort you all put in to review pull requests and ensure that the types in DefinitelyTyped really are high-quality. this is a non-obvious distinction and I'm trying to help clarify your guidelines. I'd be happy to submit a pull request to TypeScript-Handbook to distinguish these two kinds of callback functions.